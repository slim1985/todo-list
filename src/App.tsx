import { Provider } from 'react-redux';
import { store } from './store/store';
import { useAuth } from './hooks/useAuth';
import { LoginForm } from './components/LoginFormComponent/LoginForm';
import { AppContainer } from './components/AppContainerComponent/AppContainer';

export default function App(): JSX.Element {
    const { isAuthenticated, user, authenticate, signOut } = useAuth();

    return (
        <Provider store={store}>
            {isAuthenticated ? (
                <AppContainer user={user} signOut={signOut} />
            ) : (
                <LoginForm authenticate={authenticate} />
            )}
        </Provider>
    );
}
