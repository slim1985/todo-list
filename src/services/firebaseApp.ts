import { initializeApp } from 'firebase/app';

const firebaseConfig = {
    apiKey: 'AIzaSyCiA6n6d7-fjZyhJXHTX06WX7RMrWdlTpM',
    authDomain: 'ssulimov-todo.firebaseapp.com',
    projectId: 'ssulimov-todo',
    storageBucket: 'ssulimov-todo.appspot.com',
    messagingSenderId: '177684971035',
    appId: '1:177684971035:web:55d9d0a48e9d92d20b42be',
    measurementId: 'G-9Z8DMNW1ZE',
};

export const firebaseApp = initializeApp(firebaseConfig);
