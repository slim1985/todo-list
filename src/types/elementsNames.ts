export enum ElementsNames {
    STATUS_SELECT = 'status-select',
    TITLE_TEXT_AREA = 'title-textarea',
    DESCRIPTION_TEXT_AREA = 'description-textarea',
}
