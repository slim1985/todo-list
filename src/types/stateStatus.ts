export enum StateStatus {
    LOADING = 'loading',
    IDLE = 'idle',
    FAILED = 'failed',
}
