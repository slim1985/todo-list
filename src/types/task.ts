export interface Task {
    id: string;
    title: string;
    description: string;
    state: TaskStates;
}

export enum TaskStates {
    new = 'new',
    active = 'active',
    completed = 'completed',
}
