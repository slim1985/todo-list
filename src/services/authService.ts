import {
    Auth,
    connectAuthEmulator,
    getAuth,
    GoogleAuthProvider,
    signInWithPopup,
    signOut,
} from 'firebase/auth';
import { firebaseApp } from '../services/firebaseApp';
import { modes } from '../utils/constants';

export class AuthService {
    public get firebaseAuth(): Auth {
        return this._firebaseAuth;
    }

    private _firebaseAuth: Auth;
    private googleAuthProvider: GoogleAuthProvider;

    public get currentUserId(): string | null {
        return this.firebaseAuth.currentUser?.uid || null;
    }

    constructor() {
        this._firebaseAuth = getAuth(firebaseApp);
        if (process.env.NODE_ENV === modes.DEVELOPMENT) {
            connectAuthEmulator(this._firebaseAuth, 'http://127.0.0.1:9099', {
                disableWarnings: true,
            });
        }

        this.googleAuthProvider = new GoogleAuthProvider();
    }

    authenticate(): void {
        if (this.firebaseAuth.currentUser) {
            console.warn('Already authenticated. Please sign out first.');
            return;
        }

        signInWithPopup(this.firebaseAuth, this.googleAuthProvider).catch(
            (error) => {
                console.error(`Authentication error: ${error}`);
            },
        );
    }

    signOut(): void {
        if (!this.firebaseAuth.currentUser) {
            console.warn('Already signed out. Please authenticate first.');
            return;
        }

        signOut(this.firebaseAuth).catch((error) => {
            console.error(`Sign-out error: ${error}`);
        });
    }
}

export const authService = new AuthService();
