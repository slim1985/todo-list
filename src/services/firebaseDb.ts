import {
    getFirestore,
    collection,
    getDocs,
    Firestore,
    addDoc,
    getDoc,
    deleteDoc,
    doc,
    setDoc,
} from 'firebase/firestore/lite';
import { firebaseApp } from './firebaseApp';
import { Task } from '../types/task';

export class FirebaseDb {
    private firestore: Firestore;

    public constructor() {
        this.firestore = getFirestore(firebaseApp);
    }

    public async getTasks(): Promise<Task[]> {
        const tasksSnapshot = await getDocs(collection(this.firestore, 'todo'));
        const tasks = tasksSnapshot.docs.map((doc) => {
            const data = doc.data();
            return {
                id: doc.id,
                title: data.title,
                description: data.description,
                status: data.status,
            };
        });

        return tasks;
    }

    public async createTask(task: Task): Promise<Task> {
        const docRef = await addDoc(collection(this.firestore, 'todo'), {
            title: task.title,
            description: task.description,
            status: task.status,
        });
        const newDoc = await getDoc(docRef);
        const newTask: Task = {
            id: docRef.id,
            title: newDoc.data()!.title,
            description: newDoc.data()!.description,
            status: newDoc.data()!.status,
        };

        return newTask;
    }

    public async updateTask(task: Task): Promise<Task> {
        const docRef = doc(this.firestore, 'todo', task.id);
        await setDoc(docRef, {
            title: task.title,
            description: task.description,
            status: task.status,
        });

        const docData = await getDoc(docRef);
        const updatedTask: Task = {
            id: docData.id,
            title: docData.data()!.title,
            description: docData.data()!.description,
            status: docData.data()!.status,
        };

        return updatedTask;
    }

    public async deleteTaskAsync(taskId: string): Promise<string> {
        const docRef = doc(collection(this.firestore, 'todo'), taskId);
        await deleteDoc(docRef);
        return taskId;
    }
}

export const firebaseDb = new FirebaseDb();
