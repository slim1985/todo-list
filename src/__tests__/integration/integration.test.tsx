import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { getAuth, User } from 'firebase/auth';
import App from '../../App';
import { TaskData, TaskStates } from '../../types/task';
import { FirebaseDocument } from '../../services/firebaseDb';
import { ElementsNames } from '../../types/elementsNames';

jest.mock('../../services/firebaseDb', () => {
    const tasks: FirebaseDocument[] = [
        {
            id: '1',
            data: {
                title: 'Task 1',
                description: 'Description 1',
                status: '0',
            },
        },
        {
            id: '2',
            data: {
                title: 'Task 2',
                description: 'Description 2',
                status: '1',
            },
        },
    ];

    return {
        firebaseDb: {
            getManyAsync: jest.fn(() => {
                return tasks;
            }),
            createOneAsync: jest.fn(
                (collectionName: string, data: TaskData) => {
                    let lastId: number = 1;
                    let found = false;
                    while (!found) {
                        found = true;
                        tasks.forEach((task) => {
                            const id = parseInt(task.id);
                            if (id === lastId) {
                                lastId++;
                                found = false;
                            }
                        });
                    }

                    const length = tasks.push({
                        id: lastId.toString(),
                        data: data,
                    });

                    return tasks[length - 1];
                },
            ),
            updateOneAsync: jest.fn(
                (collectionName: string, docId: string, data: TaskData) => {
                    const taskIndex = tasks.findIndex(
                        (task) => task.id === docId,
                    );
                    tasks[taskIndex].data = data;

                    return tasks[taskIndex];
                },
            ),
            deleteOneAsync: jest.fn((collectionName: string, docId: string) => {
                const taskIndex = tasks.findIndex((task) => task.id === docId);
                tasks.splice(taskIndex, 1);

                return docId;
            }),
        },
    };
});

jest.mock('firebase/auth', () => {
    type callbackType = (user: User | null) => void;

    const authInstance = {
        currentUser: null as User | null,
        updateCurrentUser: jest.fn((user: User | null) => {
            authInstance.currentUser = user;
        }),
    };

    const authCallBacks: callbackType[] = [];

    return {
        getAuth: jest.fn(() => authInstance),

        onAuthStateChanged: jest.fn((auth, callback) => {
            authCallBacks.push(callback);
        }),

        signInWithPopup: jest.fn(() => {
            const user = {
                uid: '1',
                displayName: 'Test',
                email: 'test@test.com',
            } as User;
            authInstance.updateCurrentUser(user);
            authCallBacks.forEach((callback) => {
                callback(user);
            });

            return Promise.resolve();
        }),

        signOut: jest.fn(() => {
            authInstance.updateCurrentUser(null);
            authCallBacks.forEach((callback) => {
                callback(null);
            });

            return Promise.resolve();
        }),

        GoogleAuthProvider: jest.fn(() => {
            return {};
        }),
    };
});

jest.mock('../../services/firebaseApp', () => {
    return {
        firebaseApp: {},
    };
});

describe('Integration tests', () => {
    it('integration test', async () => {
        // Arrange
        getAuth().updateCurrentUser(null);

        // Act 1. Render the authentication page.
        const { asFragment, getByText, getByTestId } = render(<App />);
        expect(asFragment()).toMatchSnapshot('1. Authentication page');

        // Act 2. Sign-in by Google.
        await waitFor(() => getByText('Sign-in by Google').click());
        expect(asFragment()).toMatchSnapshot('2. Tasks page');

        // Act 3. Open task.
        await waitFor(() => getByText('Task 1').click());
        expect(asFragment()).toMatchSnapshot('3. Task page');

        // Act 4. Update task.
        await waitFor(() => {
            const selctElement = getByTestId(
                ElementsNames.STATUS_SELECT,
            ) as HTMLSelectElement;
            fireEvent.change(selctElement, {
                target: { value: TaskStates.COMPLETED },
            });

            const titleTextArea = getByTestId(
                ElementsNames.TITLE_TEXT_AREA,
            ) as HTMLTextAreaElement;
            fireEvent.change(titleTextArea, {
                target: { value: 'Task 1 updated' },
            });

            const descriptionTextArea = getByTestId(
                ElementsNames.DESCRIPTION_TEXT_AREA,
            ) as HTMLTextAreaElement;
            fireEvent.change(descriptionTextArea, {
                target: { value: 'Description 1 updated' },
            });
        });
        expect(asFragment()).toMatchSnapshot('4. Task page changed');

        // Act 5. Save updated task.
        await waitFor(() => getByText('Save').click());
        expect(asFragment()).toMatchSnapshot('5. Task page saved');

        // Act 6. Create and save task.
        await waitFor(() => getByText('Create').click());
        await waitFor(() => getByText('Save').click());
        expect(asFragment()).toMatchSnapshot('6. Task page created and saved');

        // Act 7. Open and close task form.
        await waitFor(() => getByText('Title').click());
        await waitFor(() => getByText('Cancel').click());
        expect(asFragment()).toMatchSnapshot('7. Task page open and close');

        // Act 8. Delete task.
        await waitFor(() => getByText('Title').click());
        await waitFor(() => getByText('Delete').click());
        expect(asFragment()).toMatchSnapshot('8. Delete task');

        // Act 9. Sign-out.
        await waitFor(() => getByText('Sign Out').click());
        expect(asFragment()).toMatchSnapshot('8. Sign out');
    });
});
