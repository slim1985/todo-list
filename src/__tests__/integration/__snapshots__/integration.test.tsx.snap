// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests integration test: 1. Authentication page 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-center text-2xl retina:text-7xl mt-7 retina:mt-16 text-gray-600"
    >
      Please Sign-In
    </h1>
    <div
      class="flex w-full justify-center mt-14 retina:mt-28"
    >
      <button
        class="border-solid border-gray-400 border-2 retina:border-4 py-2 retina:py-6 px-2 retina:px-6 rounded-lg retina:rounded-3xl retina:text-7xl"
      >
        Sign-in by Google
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 2. Tasks page 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            test@test.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 3. Task page 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
        </div>
      </div>
      <div
        class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
      >
        <div
          class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
        >
          <select
            class="text-3xl retina:text-8xl m-1 retina:m-3"
            data-testid="status-select"
            name="status"
          >
            <option
              value="0"
            >
              New
            </option>
            <option
              value="1"
            >
              Active
            </option>
            <option
              value="2"
            >
              Completed
            </option>
          </select>
          <textarea
            class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="title-textarea"
            name="title"
          >
            Task 1
          </textarea>
          <textarea
            class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="description-textarea"
            name="description"
          >
            Description 1
          </textarea>
          <div
            class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
          >
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Save
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Cancel
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 4. Task page changed 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
        </div>
      </div>
      <div
        class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
      >
        <div
          class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
        >
          <select
            class="text-3xl retina:text-8xl m-1 retina:m-3"
            data-testid="status-select"
            name="status"
          >
            <option
              value="0"
            >
              New
            </option>
            <option
              value="1"
            >
              Active
            </option>
            <option
              value="2"
            >
              Completed
            </option>
          </select>
          <textarea
            class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="title-textarea"
            name="title"
          >
            Task 1 updated
          </textarea>
          <textarea
            class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="description-textarea"
            name="description"
          >
            Description 1 updated
          </textarea>
          <div
            class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
          >
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Save
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Cancel
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 5. Task page saved 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            test@test.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1 updated
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1 updated
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 6. Task page created and saved 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            test@test.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1 updated
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1 updated
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Title
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 7. Task page open and close 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            test@test.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1 updated
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1 updated
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Title
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 8. Delete task 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            test@test.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1 updated
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 1 updated
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Description 2
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Integration tests integration test: 8. Sign out 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-center text-2xl retina:text-7xl mt-7 retina:mt-16 text-gray-600"
    >
      Please Sign-In
    </h1>
    <div
      class="flex w-full justify-center mt-14 retina:mt-28"
    >
      <button
        class="border-solid border-gray-400 border-2 retina:border-4 py-2 retina:py-6 px-2 retina:px-6 rounded-lg retina:rounded-3xl retina:text-7xl"
      >
        Sign-in by Google
      </button>
    </div>
  </div>
</DocumentFragment>
`;
